### PR TITLE
Added multi embeddings

### PR DESCRIPTION
**Pull request by mjason98**

### CODE REVIEW

File: reverse_image_search_module/dic_to_list.py [hunk 1]
```diff
@@ -1,20 +1,49 @@
...
+    if len(value_shapes) != 1:
+        raise ValueError("All values in the dictionary must have the same shape.")
+    keys_array = np.array(keys)
+    values_array = np.array(values)
+    np.save(keys_output_file, keys_array)
+    np.save(values_output_file, values_array)
+
+
+def multi_embedding():
+    input_file = 'data/embeddings_full_multi.pkl'
```
Comment (aguschin) [inline reverse_image_search_module/dic_to_list.py:26]: I’d make these function arguments (without default values) and pass the right values when you call this function.

File: reverse_image_search_module/search_image.py [hunk 2]
```diff
@@ -5,11 +5,13 @@
 import pandas as pd
 import torchvision.transforms as transforms
 from PIL import Image
+from collections import Counter
+
 
 # Load all embeddings from the .npy filessssssssssssssssssss
-all_embeddings = np.load('./data/embeddings.npy')
+all_embeddings = np.load('./data/embeddings_multi.npy')
```
Comment (aguschin) [inline reverse_image_search_module/search_image.py:12]: I think we have a little bit too many hardcoded paths in the project now. Let’s set them up in `.env.example` and `.env`, WDYT?

Comment (aguschin) [reply reverse_image_search_module/search_image.py:12]: And I’d implement some project-wide configuration switch: you either use multicropping or single cropping. If it make sense to use it in several files, then it would make things simpler to change - if you change it in one place, all other places are work in sync, so you won’t encounter problems like if you forgot to turn it on in all places. Same logic as with these paths I mentioned in the previous comment. WDYT?

File: reverse_image_search_module/resnet18.py [hunk 3]
```diff
@@ -81,7 +83,86 @@ def extract_and_save_embeddings(input_folder, output_file):
...
-    extract_and_save_embeddings(input_folder, output_file)
+    IMAGE_FOLDER = 'data'
+    
+    if not os.path.isdir(IMAGE_FOLDER):
+        os.mkdir(IMAGE_FOLDER)
+    
+    input_folder = 'data/img/full'
+    output_file = f'{IMAGE_FOLDER}/embeddings_full_multi.pkl'
+    
+    # extract_and_save_embeddings(input_folder, output_file)
```
Comment (aguschin) [inline reverse_image_search_module/resnet18.py:167]: This is the place where we should use the switch (single/multi) I’m taking about in the other comment.

File: reverse_image_search_module/search_image.py [hunk 4]
```diff
@@ -81,6 +83,55 @@ def find_index_from_image(img, n):
...
 
+def find_repeating_index(idx, dist):
+    counter = {}
+
+    for ide, distance in zip(idx, dist):
+        if counter.get(ide, None) is None:
+            record = [distance, 1]
+        else:
+            record = counter[ide]
+            record[0] += 1.0 /distance
```
Comment (aguschin) [inline reverse_image_search_module/search_image.py:94]: Distance here is the cosine similarity, right? Between -1 and 1

File: reverse_image_search_module/test/test_multicrop.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+            _ = image.resize((255, 255))
+        except Exception as ex:
+            mylogger.info(f'image {_im} load error')
+            mylogger.info(str(ex))
+            continue
+        
+        
+        idxs, dists = [], []
+
+        for x_ini, y_ini, x_end, y_end in random_cropping(255,255, 200, K_CROPPING):
```
Comment (aguschin) [inline reverse_image_search_module/test/test_multicrop.py:48]: Random croppings here may lead to non-reproduced behavior: like if I’m submitting the same image to art bot twice and getting different results. Pros: if the image wasn’;t found once, it can be found on the second run (works in other direction as well). Cons: can be hard to debug a specific situation and see why we got some results - this won’t be reproducible.

File: reverse_image_search_module/test/test_multicrop.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+            
+            idxs.append(idx)
+            dists.append(dist)
+        
+        selected_idx, selected_dist = max(zip(idxs, dists), key=lambda x: x[1])
+
+        file_name = find_file_name(selected_idx)
+
+        if file_name == _im:
+            positive += 1
```
Comment (aguschin) [inline reverse_image_search_module/test/test_multicrop.py:68]: This may be a wrong look at this case, but logic like this (calculating the metric) is usually found one level higher. If this function could just find the closest and return the list of closest, you can then perform calculation of the metric. Please ignore if you feel this is nitpicking :)

Comment (aguschin) [reply reverse_image_search_module/test/test_multicrop.py:68]: The other solution: rename the function. “Get image” in the name means you’re going to have some images maybe? But not the number which is a metric value. This should remove ambiguity for me.

Comment (aguschin) [reply reverse_image_search_module/test/test_multicrop.py:68]: Upon looking at the next function, I think renaming is better since metric calculation is involved process and it will be unnecessary complicated to return all images, thresholds, etc. The only pro in that case is if you want to make these functions more flexible: like you want to score different approaches (1:1, m:1, etc) against different ways of calculating the match, I.e. the image you find (just max cosine, harmonic mean, etc).

**Review by aguschin — CHANGES_REQUESTED** [review]
Hey, there are some things we can improve upon I believe. WDYT?

File: reverse_image_search_module/search_image.py (hunk 2 above)
Comment (mjason98) [reply reverse_image_search_module/search_image.py]: Resolved in commit 05a576b :smile:

File: reverse_image_search_module/resnet18.py (hunk 3 above)
Comment (mjason98) [reply reverse_image_search_module/resnet18.py]: resolved in commit 05a576b :muscle:

File: reverse_image_search_module/search_image.py (hunk 4 above)
Comment (mjason98) [reply reverse_image_search_module/search_image.py]: I remove the function :) it was ussless at the end

File: reverse_image_search_module/dic_to_list.py (hunk 1 above)
Comment (mjason98) [reply reverse_image_search_module/dic_to_list.py]: Fixed in commit b17be5b :broom:

File: reverse_image_search_module/test/test_multicrop.py (hunk 5 above)
Comment (mjason98) [reply reverse_image_search_module/test/test_multicrop.py]: solved in commit f07d64c :)

File: reverse_image_search_module/test/test_multicrop.py (hunk 6 above)
Comment (mjason98) [reply reverse_image_search_module/test/test_multicrop.py]: Resolved in commit 91e1a92 :)

File: reverse_image_search_module/resnet18.py [hunk 7]
```diff
@@ -81,7 +86,89 @@ def extract_and_save_embeddings(input_folder, output_file):
...
+
+            vectors= []
+
+            for x, y, x_end, y_end in gen_multi_cropping(image.width, image.height, k=k, min_size_random=128):
+                croped = image.crop((x, y, x_end, y_end))
+                feature_vector = img2vec.getVectors(croped)
+
+                vectors.append(feature_vector)
+                # print(feature_vector.shape)
+                # assert False
```
Comment (aguschin) [inline reverse_image_search_module/resnet18.py:144]:
```suggestion
```

File: reverse_image_search_module/dic_to_list.py [hunk 8]
```diff
@@ -1,20 +1,53 @@
...
+        raise ValueError("All values in the dictionary must have the same shape.")
+    keys_array = np.array(keys)
+    values_array = np.array(values)
+    np.save(keys_output_file, keys_array)
+    np.save(values_output_file, values_array)
+
+
+def multi_embedding(input_file = 'data/embeddings_full_multi.pkl', 
+                    values_output_file = 'data/embeddings_multi.npy',
+                    keys_output_file = 'data/file_names_multi.npy'):
```
Comment (aguschin) [inline reverse_image_search_module/dic_to_list.py:30]: I'd move these hardcoded paths to `.env` since you're reusing them several times in the project, but if you disagree let's keep it as is.

**Review by aguschin — APPROVED** [review]
LGTM except for the one comment I left. Great job @mjason98!
